src/auto_renewal.cairo

```diff
@@ -0,0 +1,242 @@
+            ref self: ContractState,
+            domain: array::Array::<felt252>,
+            renewer: array::Array::<starknet::ContractAddress>,
+            limit_price: array::Array::<u256>,
+        ) {
+            assert(domain.len() == renewer.len(), 'Domain & renewer mismatch len');
+            assert(domain.len() == limit_price.len(), 'Domain & price mismatch len');
+
+            let mut domain = domain;
+            let mut renewer = renewer;
+            let mut limit_price = limit_price;
+
+            loop {
+                if domain.len() == 0 {
+                    break;
+                }
+                let _domain = domain.pop_front().expect('pop_front error');
+                let _renewer = renewer.pop_front().expect('pop_front error');
+                let _limit_price = limit_price.pop_front().expect('pop_front error');
+                self._renew(_domain, _renewer, _limit_price);
+            }
+        }
+    }
+
+    //
+    // Internals
+    //
+
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        fn _renew(
+            ref self: ContractState,
+            root_domain: felt252,
+            renewer: ContractAddress,
+            limit_price: u256,
+        ) {
+            let naming = self.naming_contract.read();
+            let can_renew = self._is_renewing.read((renewer, root_domain, limit_price));
+            assert(can_renew == 1, 'Renewal not toggled for domain');
+
+            // Check domain has not been renewed yet this year
+            let block_timestamp: felt252 = get_block_timestamp().into();
+            let last_renewed = self.last_renewal.read((renewer, root_domain));
+            let elapsed: u256 = u256_from_felt252(block_timestamp - last_renewed);
+            let max = u256_from_felt252(86400 * 364);
+            assert(elapsed > max, 'Domain already renewed');
+
+            // Check domain is set to expire within a month
+            let mut domain_arr = ArrayTrait::<felt252>::new();
+            domain_arr.append(root_domain);
+            let expiry = INamingDispatcher {
+                contract_address: naming
+            }.domain_to_expiry(domain_arr);
+            assert(
+                u256_from_felt252(expiry) <= u256_from_felt252(block_timestamp)
+                    + u256_from_felt252(86400 * 30),
+                'Domain not set to expire'
+            );
+
+            // Check renew price for domain is lower or equal to limit price
+            let pricing = self.pricing_contract.read();
+            let (erc20, renewal_price) = IPricingDispatcher {
+                contract_address: pricing
+            }.compute_renew_price(root_domain, 365);
+            assert(renewal_price <= limit_price, 'Renewal price > limit price');
+
+            // Renew domain
+            let contract = get_contract_address();
+            IERC20Dispatcher {
+                contract_address: erc20
+            }.transferFrom(renewer, contract, renewal_price);
```

Comment:
Here, before calling .renew, we should call Eth.allow(naming, limit_price), this will ensure that the naming contract can't spend more than the limit set by the user.